### PR TITLE
[Frontend] JAX tracer supporting user-defined tape transformation

### PR DESCRIPTION
**Pull request by grwlf**
In this PR we aim at making it possible to apply user-defined transformations to a tape containing Catalyst quantum operators. The main difficulty is related to the fact that in Catalyst we only trace programs rather than execute them so the operations on tape: (1) are filled with abstract parameters, (2) include control flow operations containing nested tapes.

The approach we are working on would essentially split the tracing process into classical and quantum parts. In order to do so, we store JAX tracing contexts and re-enter them as needed. We described the idea and the associated risks in the [internal document](https://github.com/XanaduAI/sergei-environment/blob/master/src/python_side_transformations/pst.md).

Tasks:
- [x] Prototyping is done
- [x] Core tracing framework is implemented
- [x] Catalyst extensions are ported to the new tracing framework
  - [x] Control flow ops (For-loops, WhileLoops, Conditionals, Adjoint)
   - [x] CustomOps
   - [x] QubitUnitary
  - [x] Mid-circuit measurement
  - [x] Measurements (Sample, ExpVal, Counts, Probs, State)
  - [x] Observables (Compbasis, Paulis, Hadamard, Hermitian,Tensor,Hamiltonian)
  - [x] Gradient primitives (Grad, JVP, VJP)
- [x] All kinds of execution pipelines are supported
  - [x] Both qjit and qnode
  - [x] Just qjit (no quantum parts)
  - [x] Just qnode (interpretation) 
 - [x] Optimizations are ported
   - [x] Qubit insert/extract state preservation (quantum register tracing optimisation)
   - [x] Differentiation-specific procedures (JVP/VJP/Grad) are reviewed and ported
- [x] All tests pass locally 
- [x] Latest version of Catalyst is merged into the PR
  - [x] Obvious merge conflicts are solves
  - [x] Pytests are fixed
  - [x] Littests are fixed
- [ ] Code is clean
  - [x] Temporary and unused source files are removed
  - [x] JAX internals are collected in a separate utility file
  - [x] Tracing context classes are unified
  - [x] Pylint is happy
  - [x] Docstrings are ported or created
  - [x] CodeFactor is happy
  - [x] Codecov is happy
  - [ ] Remaining FIXMEs are fixed

[sc-41345]

### CODE REVIEW

**Comment by grwlf** [comment]
Hi colleagues, I think this PR approaches the ready state. Some docstrings are not here yet, but I think I can ask for technical review already. @dime10 @erick-xanadu

File: frontend/catalyst/jax_tracer.py
Comment (dime10) [inline frontend/catalyst/jax_tracer.py:1]: It appears that a lot of definitions were moved from the `pennylane_extensions` module to here. Was this a conscious choice or a holdover from prototyping?
I'm not necessarily against reorganization if it improves clarity or encapsulation, but I don't see the benefit in this instance yet. (The tracing module looks much heavier now than it might need to be.)

For some context on the original module organization:
The `jax_tracer` module should mostly be concerned with the process of tracing in general, rather than defining objects that might be traced. The `jax_primitives` module is used to define new objects on the jax side, while `pennylane_extensions` contains definitions for new objects on the PL side. In particular, the latter would refer to anything that inherits from `pennylane.operation.Operation`.

File: frontend/test/pytest/test_mid_circuit_measurement.py [hunk 1]
```diff
@@ -47,6 +47,18 @@ def circuit():
...
+    def test_invalid_arguments(self, backend):
+        """Test invalid arguments exception."""
+
+        @qml.qnode(qml.device(backend, wires=2))
+        def circuit():
+            qml.RX(0.0, wires=0)
+            m = measure(wires=[1, 2])
+            return m
+
+        with pytest.raises(TypeError, match="One classical argument \(a wire\) is expected"):
```
Comment (dime10) [inline frontend/test/pytest/test_mid_circuit_measurement.py:59]:
```suggestion
        with pytest.raises(TypeError, match=r"One classical argument \(a wire\) is expected"):
```

File: frontend/test/pytest/test_compiler.py [hunk 2]
```diff
@@ -422,6 +423,17 @@ def get_output_filename(infile):
...
+    def test_qfunc_no_device(self):
+        """Test qfunc with device=None"""
+
+        @qjit()
+        @qfunc(1, device=None)
+        def circuit():
+            qml.PauliX(wires=0)
+            return measure(wires=0)
+
+        assert circuit()
```
Comment (dime10) [inline frontend/test/pytest/test_compiler.py:435]: Is this something we allow? 😅
We should probably delete the `QFunc` class constructor actually, I don't think the class is intended to be used anywhere anymore, except to patch the `QNode.__call__` method.

Having said that, if we keep this for now I don't think the tests thematically fits here. `test_qnode.py` might be the closest, or simply in its own `test_qfunc.py` :)

File: frontend/test/lit/test_multi_qubit_gates.py [hunk 3]
```diff
@@ -55,7 +55,7 @@ def circuit():
         ]
     )
     # CHECK: {{%.+}} = "quantum.unitary"({{%.+}}, {{%.+}}, {{%.+}}) : (tensor<4x4xcomplex<f64>>, !quantum.bit, !quantum.bit) -> (!quantum.bit, !quantum.bit)
-    qml.QubitUnitary(U2, wires=[1, 2])
+    qml.QubitUnitary(U2, wires=[0, 2])
```
Comment (dime10) [inline frontend/test/lit/test_multi_qubit_gates.py:58]: No change required, just wondering if there was a bug in the test before?

File: frontend/test/lit/test_if_else.py [hunk 4]
```diff
@@ -24,9 +24,9 @@
 @qjit(target="mlir")
 @qml.qnode(qml.device("lightning.qubit", wires=1))
 def circuit(n: int):
-    # CHECK-DAG:   [[qreg_0:%[a-zA-Z0-9_]+]] = "quantum.alloc"
     # CHECK-DAG:   [[c5:%[a-zA-Z0-9_]+]] = stablehlo.constant dense<5> : tensor<i64>
     # CHECK:       [[b_t:%[a-zA-Z0-9_]+]] = stablehlo.compare  LE, %arg0, [[c5]], SIGNED : (tensor<i64>, tensor<i64>) -> tensor<i1>
+    # CHECK-DAG:   [[qreg_0:%[a-zA-Z0-9_]+]] = "quantum.alloc"
```
Comment (dime10) [inline frontend/test/lit/test_if_else.py:29]: This implies the quantum allocation instruction is not guaranteed to be generated first anymore, right?

File: frontend/test/lit/test_decomposition.py [hunk 5]
```diff
@@ -257,26 +257,26 @@ def test_decompose_singleexcitationplus():
...
         # CHECK: [[s0q0:%.+]] = "quantum.custom"({{%.+}}) {gate_name = "PauliX"
         # CHECK-NOT: name = "SingleExcitationPlus"
+        # CHECK: [[a_scalar_tensor_float_2:%.+]] = stablehlo.constant dense<2.{{[0]+}}e+00>
+        # CHECK-NOT: name = "SingleExcitationPlus"
+        # CHECK: [[a_theta_div_2:%.+]] = stablehlo.divide %arg0, [[a_scalar_tensor_float_2]]
+        # CHECK-NOT: name = "SingleExcitationPlus"
         # CHECK: [[a_theta_div_2_scalar:%.+]] = "tensor.extract"([[a_theta_div_2]])
         # CHECK-NOT: name = "SingleExcitationPlus"
-        # CHECK: [[s1:%.+]]:2 = "quantum.custom"([[a_theta_div_2_scalar]], [[s0q0]], [[s0q1]]) {gate_name = "ControlledPhaseShift"
+        # CHECK: [[s1:%.+]]:2 = "quantum.custom"([[a_theta_div_2_scalar]], [[s0q1]], [[s0q0]]) {gate_name = "ControlledPhaseShift"
```
Comment (dime10) [inline frontend/test/lit/test_decomposition.py:270]: Also curious about this change, the order of the qubits was swapped with no change in the example?

File: frontend/test/pytest/test_operations.py [hunk 6]
```diff
@@ -143,5 +145,48 @@ def circuit(x: float, y: float):
...
+                qml.RZ(np.pi, wires=0)
+                return 1
+
+            cond_fn()
+
+        for term in ["ForLoop", "WhileLoop", "Cond", "RX", "RY", "RZ"]:
+            assert term in str(quantum_tape.operations)
+        for op in quantum_tape.operations:
+            if isinstance(op, (ForLoop, WhileLoop, Cond)):
+                assert has_nested_tapes(op)
```
Comment (dime10) [inline frontend/test/pytest/test_operations.py:185]: It might be good to explicitly test what you are expecting here, so that we don't need to rely on the correctness of `has_nested_tapes` by itself (although it can still be called for verification).

**Review by dime10 — COMMENTED** [review]
Just had a first look from a high level and at the tests, will tack on a review of the main functionality in a bit :)

One thing I noticed is that we removed all the unit tests associated with the old tracing infra, but really add any new ones. Do you think there are some useful ones that we could add?

File: frontend/test/lit/test_multi_qubit_gates.py (hunk 3 above)
Comment (grwlf) [reply frontend/test/lit/test_multi_qubit_gates.py:58]: I think we could call it an error in test, but it depends on our assumptions regarding the hardware. In the original version of the test the first quantum op affected wire 0 and the second one affected wires 1 and 2 (and btw we are measuring wire 0 only). Given the jax sorting, there is no good way to force the final program to match the littest's expectations. In this particular case the algorithm placed the second op before the first one. So I changed the wire to make the order to really match the lit-rules we have here.

Comment (grwlf) [reply frontend/test/lit/test_multi_qubit_gates.py:58]: I think we could call it an error in test, but it depends on our assumptions regarding the hardware. In the original version of the test the first quantum op affected wire 0 and the second one affected wires 1 and 2 (and btw we are measuring wire 0 only). Given the jax sorting, there is no good way to force the final program to match the littest's expectations. In this particular case the algorithm placed the second op before the first one. So I changed the wire to make the order to really match the lit-rules we have here.

File: frontend/test/pytest/test_compiler.py (hunk 2 above)
Comment (grwlf) [reply frontend/test/pytest/test_compiler.py:435]: Hmm, that must be one of the tests I added to increase the coverage. I saw in code that we must be expecting this. Probably just removing it would be better if you are in doubts.

File: frontend/test/lit/test_decomposition.py (hunk 5 above)
Comment (grwlf) [reply frontend/test/lit/test_decomposition.py:270]: That is the consequence of jax-sorting. The algorithm is deterministic, but the particular order of the equations might differ from test case to test case. Most of the tests just match the expectations but some of them don't. In these cases I needed to either change the algorithm or to change the test. Since I couldn't see clear rules to implement, I decided to change the test

Comment (grwlf) [reply frontend/test/lit/test_decomposition.py:270]: It is the consequence of jax-sorting. The sorting algorithm is deterministic, but the particular order of the equations might differ from test case to test case. Most of the tests just match the expectations but some of them don't. In these cases I needed to either change the algorithm or to change the test.

The point here is that we do not lit-check every line. The order of initial qubit-producing instructions must have been changed by the sorting (something we didn't capture in this test) so mlir assigned names in a different way.

File: frontend/test/lit/test_multi_qubit_gates.py (hunk 3 above)
Comment (erick-xanadu) [reply frontend/test/lit/test_multi_qubit_gates.py:58]: So you are saying essentially that `qml.QubitUnitary` is dead code and the compiler may rearrange it to be placed even after the deallocation? Or only after the measurement?

Comment (grwlf) [reply frontend/test/lit/test_multi_qubit_gates.py:58]: The deallocation by itself is not a problem because I made precautions to always include it as a last operation. But if we have
```
b,d = ..
a = f1(b)
c = f2(d)
use(a,c)
```
then the compiler will arrange f1 and f2 according to some deterministic but unspecified logic. Currently this does not always match the test expectations.

Comment (grwlf) [reply frontend/test/lit/test_multi_qubit_gates.py:58]: I am thinking how to specify this logic. It could be not _always_ possible due to mid-circuit measurements, but maybe we can improve it for other cases

Comment (erick-xanadu) [reply frontend/test/lit/test_multi_qubit_gates.py:58]: I think if it is dead code, it doesn't matter if it is re-arranged. We should probably optiimze it and remove it from the computation. (Potentially issue a warning if dead code is found as the user might not have intended it). But this is something beyond the scope of this PR.

File: frontend/catalyst/jax_tracer.py
Comment (dime10) [reply frontend/catalyst/jax_tracer.py:1]: Note that circular dependencies can easily be resolved by importing only the module rather than the specific member (e.g. `import pennylane` rather than `from pennylane.operations import ...`).

File: frontend/test/pytest/test_compiler.py (hunk 2 above)
Comment (dime10) [reply frontend/test/pytest/test_compiler.py:435]: > Probably just removing all this logic would be better if you are in doubts.

Exactly, I think we can remove it entirely 👍

File: frontend/test/lit/test_multi_qubit_gates.py (hunk 3 above)
Comment (dime10) [reply frontend/test/lit/test_multi_qubit_gates.py:58]: Just looking at the issue of testing the code, LIT allows you to specify checks without a particular order: `CHECK-DAG`
It can't always be used depending on the complexity of the of test, but here it should work.

I'm glad the test uncovered the question of rearranging code however. To be honest I'm not a huge fan that it is happening at such an early stage. From a functional point of view it is correct, but thinking about debugging code by looking at the jaxpr or initial mlir it is much easier to map from IR to source if the order were preserved. Once we run optimizations I think rearrangements are fair game.
Surprisingly the original tracing implementation managed to preserve the order rather well.

File: frontend/test/lit/test_decomposition.py (hunk 5 above)
Comment (dime10) [reply frontend/test/lit/test_decomposition.py:270]: Got it, thanks :)

File: frontend/test/lit/test_multi_qubit_gates.py (hunk 3 above)
Comment (grwlf) [reply frontend/test/lit/test_multi_qubit_gates.py:58]: Yes, speaking about the sorting - I believe it is correct to think that we split the original flow of instructions into two lists: classical instructions go directly to `frame.eqn` and quantum instructions go to the nested quantum tape. After that we merge the lists and at this point (in the current implementation) we indeed miss some information, e.g. line numbers to merge them precisely. I am pretty sure I can collect and preserve all the information to keep the original line order. Can do it if we want.

> Surprisingly the original tracing implementation managed to preserve the order rather well

The core problem we solve now is the list merging. The current implementation "adds" quantum tape (after the transformations!) to the already existing list of classical equation and then sort everything based on the properties of items - the variables they produce/consume. Can we think of some other way? I will think..

Comment (dime10) [reply frontend/test/lit/test_multi_qubit_gates.py:58]: Yes I think you are right. I guess the original implementation preserved the order by maintaining the sequence of values that connect the classical & quantum parts.

> line numbers - which is required for the precise merging
Maybe we could open a GitHub issue about it as a future improvement? That way we can still go ahead with the PR.

File: frontend/test/pytest/test_compiler.py (hunk 2 above)
Comment (grwlf) [reply frontend/test/pytest/test_compiler.py]: Done

File: frontend/test/pytest/test_operations.py (hunk 6 above)
Comment (grwlf) [reply frontend/test/pytest/test_operations.py:190]: I have added the explicit checks, but I kept the original call as well - I am using this test to test the `has_nested_tape` logic. Let me know if you have suggestions regarding this.

File: frontend/test/lit/test_decomposition.py (hunk 5 above)
Comment (grwlf) [reply frontend/test/lit/test_decomposition.py]: I have added the `qinst` operation to the [forced_order_primitives](https://github.com/PennyLaneAI/catalyst/blob/2a9ed99f03af5e0a449c937632ae529651f51005/frontend/catalyst/utils/jax_extras.py#L163) set. Now the sort is much more restricted and this test has only two arithmetic operations reordered. Please check!

Comment (dime10) [reply frontend/test/lit/test_decomposition.py]: Thanks 👍

File: frontend/catalyst/jax_tracer.py
Comment (dime10) [reply frontend/catalyst/jax_tracer.py:1]: Ok we decided that we can go ahead with the current version and will add a refactor to the technical roadmap that will attempt to create cleaner abstractions and separation of of concerns :)

File: frontend/test/lit/test_multi_qubit_gates.py (hunk 3 above)
Comment (grwlf) [reply frontend/test/lit/test_multi_qubit_gates.py:59]: Created this one https://github.com/PennyLaneAI/catalyst/issues/284

Comment (dime10) [reply frontend/test/lit/test_multi_qubit_gates.py:59]: Thanks :)

File: frontend/catalyst/utils/tracing.py [hunk 7]
```diff
@@ -16,45 +16,200 @@
 Tracing module.
```
Comment (dime10) [inline frontend/catalyst/utils/tracing.py:16]: Let's rename this file and module docstring to be more descriptive of what it does and to separate from other tracing related modules :)
Maybe `contexts.py`?

File: frontend/catalyst/utils/jax_extras.py [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+from jax._src.source_info_util import current as jax_current
+from jax._src.source_info_util import new_name_stack
+from jax._src.tree_util import (
+    PyTreeDef,
+    tree_flatten,
+    tree_structure,
+    tree_unflatten,
+    treedef_is_leaf,
+)
+from jax._src.util import partition_list, safe_map, unzip2, unzip3, wrap_name
```
Comment (dime10) [inline frontend/catalyst/utils/jax_extras.py:63]: Could you check which one of these is available from the public API (`jax....` rather than `jax._src...`)? That way it's easier for us to track when we use private JAX functions

File: frontend/catalyst/jax_tracer.py [hunk 9]
```diff
@@ -14,637 +14,902 @@
...
+
+        args_data, _ = tree_flatten(args)
+
+        wrapped = wrap_init(_eval_jaxpr)
+        retval = func_p.bind(wrapped, *args_data, fn=self)
+
+        return tree_unflatten(retval_tree, retval)
+
+
+class QJITDevice(qml.QubitDevice):
```
Comment (dime10) [inline frontend/catalyst/jax_tracer.py:799]: Could you move this class back to pennylane_extensions? I don't think it depends on any tracer functionality :)

**Review by dime10 — APPROVED** [review]
Looks good to me 💯

File: frontend/catalyst/jax_tracer.py (hunk 9 above)
Comment (grwlf) [reply frontend/catalyst/jax_tracer.py:645]: Done

File: frontend/catalyst/utils/tracing.py (hunk 7 above)
Comment (grwlf) [reply frontend/catalyst/utils/tracing.py]: Done

File: frontend/catalyst/utils/jax_extras.py (hunk 8 above)
Comment (grwlf) [reply frontend/catalyst/utils/jax_extras.py]: Done
